### PR TITLE
Handle case when default formtable requires default view

### LINKED ISSUE #4749: Collectors subform opens as read only and you can't change it
**Describe the bug**
The collectors subform opens as read only and expanded

**To Reproduce**

1. Go to collecting event/information 
2. Add a collector
3. See problem 

**Expected behavior**
You should be able to edit it and it should not open as read only. 

**Screenshots**

https://github.com/specify/specify7/assets/150940840/848a611e-4b27-414f-8813-868f191a4063

Please, also fill out the following information manually:

- OS: Windows 11
- Browser: Chrome
- Specify 7 Version: edge
- Database Name: lsumzmammals9823
- Collection name: mammals
- User Name: SpAdmin
[Specify 7 System Information - 2024-04-05T18_16_21.848Z.txt](https://github.com/specify/specify7/files/14889293/Specify.7.System.Information.-.2024-04-05T18_16_21.848Z.txt)

### PR DESCRIPTION
**Pull request by melton-jason**
Fixes #4749 

> Thankfully there is some consistency amongst the table-only views. Notably, every `altview` in the `altviews` of the view points to a view definition which is of type `formtable`. When reading a form definition for a subview on the frontend, if every altview points to a `formtable` viewdef in a view, we can likely fallback to using the `table.view` view for rendering the expanded definition of the subview (the collapsed view can still remain the original view).

Alternative implementation of #4760 which follows the above proposed solution from https://github.com/specify/specify7/pull/4760#issuecomment-2046022176

### Checklist

- [x] Self-review the PR after opening it to make sure the changes look good
      and self-explanatory (or properly documented)
- [ ] Add automated tests
- [x] Add relevant issue to release milestone

### Testing instructions
- In the instance, make sure there are no custom views defined for Collectors
- On the view definition for the CollectingEvent, make sure the `collectors` relationship is using the `Collectors` view (ignore this step if using the default): 
```xml
<cell type="subview" viewname="Collectors" id="collect" name="collectors" colspan="13" defaulttype="table" rows="3"/>
```
- Ensure the Collector subview correctly renders the default `Collector` view (pictured below) when the Subview is: 
  - [ ] In Subform Mode
  - [ ] In Grid Mode and the cell is expanded
  - [ ] Also ensure the correct collapsed view is shown when the Subview is in Grid Mode and the cell is collapsed
- On the CollectingEvent form definition, modify the viewname for the `collectors` relationship to be `Collector`
  - [ ] Ensure the collapsed view when the Subview is in Grid Mode is using the `Collector` view definition

- [ ] On the default Agent form, ensure the GroupPerson/Members subview appears as it did in `v7.9+`

---
**Default Collector View form:** 

Screenshot 2024-04-11 at 10 56 06 AM

**Default collapsed Collector View:** 

Screenshot 2024-04-11 at 10 58 28 AM

**Collapsed view definition for `Collector` View:**

Screenshot 2024-04-11 at 11 18 47 AM

### CODE REVIEW

**Review by melton-jason — COMMENTED** [review]
The relevant code changes which implement the solution are in the following: 

https://github.com/specify/specify7/blob/ec955a70403e3f482276273dc97c1d4864e046af/specifyweb/frontend/js_src/lib/components/FormParse/index.ts#L183-L188

I am not a fan of making the function asynchronous just for handling this case, although we need the ViewDefinition of the default table view (which is different than the view we are attempting to parse) to correctly resolve the expanded view definition. 

If it's a better solution, I have also considered extracting some form parsing logic from `resolveViewDefinition` to `fetchView` so that the correct view can be recursively fetched and altering the `ViewDefinition` type by adding an attribute for the expanded view definition for this case. 
Then, `resolveViewDefinition` can simply return the correctly resolved expanded view when the frontend wants to access the view with the `form` FormType.

File: specifyweb/frontend/js_src/lib/components/FormParse/index.ts [hunk 1]
```diff
@@ -119,63 +119,82 @@ export const fetchView = async (
...
     console.error(`No view definitions found for the ${view.name} view`);
     return undefined;
   }
 
+  const useDefaultExpanded =
+    formType === 'form' &&
+    new Set(Object.values(view.altviews).map((altView) => altView.viewdef))
+      .size === 1 &&
+    (toSimpleXmlNode(xmlToJson(viewDefinitions[view.altviews[0].viewdef]))
+      .attributes.type as Lowercase<FormType>) === 'formtable';
```
Comment (maxpatiiuk) [inline specifyweb/frontend/js_src/lib/components/FormParse/index.ts:188]: given that you only access attributes, you don't need toSimpleXmlNode here

you can also use getAttribute:

```suggestion
    getAttribute(xmlToJson(viewDefinitions[view.altviews[0].viewdef],'type') === 'formtable';
```

File: specifyweb/frontend/js_src/lib/components/FormParse/index.ts [hunk 2]
```diff
@@ -119,63 +119,82 @@ export const fetchView = async (
...
+
+  if (useDefaultExpanded)
+    return fetchView(strictGetTable(parseJavaClassName(view.class)).view).then(
+      async (viewDefinition) =>
+        viewDefinition === undefined
```
Comment (maxpatiiuk) [inline specifyweb/frontend/js_src/lib/components/FormParse/index.ts:193]: if this is undefined, then maybe we want to just let the function go along with it's default behavior?

```tsx
  if (useDefaultExpanded) {
    const viewDefiniton = await fetchView(strictGetTable(parseJavaClassName(view.class)).view);
    if(typeof viewDefinition === 'object')
      return resolveViewDefinition(viewDefinition, formType, mode)
  }
```

**Review by maxpatiiuk — APPROVED** [review]
sad to see more complications being added to an already complex functions, but this seems like the best way forward

could you also please add tests for this new logic (i.e could add new tests for resolveViewDefinition, or extract the useDefaultExpanded computation logic into a pure function and test that using unit tests)

**Comment by melton-jason** [comment]
> If it's a better solution, I have also considered extracting some form parsing logic from `resolveViewDefinition` to `fetchView` so that the correct view can be recursively fetched and altering the `ViewDefinition` type by adding an attribute for the expanded view definition for this case. Then, `resolveViewDefinition` can simply return the correctly resolved expanded view when the frontend wants to access the view with the `form` FormType.

I have gone ahead and implemented a (hopefully) cleaner solution generally following the above proposal which is less bug prone. 
Now when fetching the view from the backend, if we are fetching one of the table-only views, we fetch the default view for the table and add the non-formtable altviews and viewdefs to the table-only view. 
When the frontend attempts to access the 'expanded' form definition for the altered view, it can then correctly resolve the expanded view definition. 

Essentially, we are making every view follow the pattern of other views (like the default `Preparation` view in https://github.com/specify/specify7/pull/4760#issuecomment-2046022176)

**Review by emenslin — APPROVED** [review]
### Testing instructions
- In the instance, make sure there are no custom views defined for Collectors
- On the view definition for the CollectingEvent, make sure the `collectors` relationship is using the `Collectors` view (ignore this step if using the default): 
```xml
<cell type="subview" viewname="Collectors" id="collect" name="collectors" colspan="13" defaulttype="table" rows="3"/>
```
- Ensure the Collector subview correctly renders the default `Collector` view (pictured below) when the Subview is: 
  - [x] In Subform Mode
  - [x] In Grid Mode and the cell is expanded
  - [ ] Also ensure the correct collapsed view is shown when the Subview is in Grid Mode and the cell is collapsed
- On the CollectingEvent form definition, modify the viewname for the `collectors` relationship to be `Collector`
  - [x] Ensure the collapsed view when the Subview is in Grid Mode is using the `Collector` view definition

- [ ] On the default Agent form, ensure the GroupPerson/Members subview appears as it did in `v7.9+`

Group person subform opens as right aligned and in grid mode it originally opens as collapsed and read only and it should open expanded. 

https://github.com/specify/specify7/assets/150940840/333e6ab8-8ceb-4042-943c-f414e4eadc73

**Review by emenslin — CHANGES_REQUESTED** [review]
### Testing instructions
- In the instance, make sure there are no custom views defined for Collectors
- On the view definition for the CollectingEvent, make sure the `collectors` relationship is using the `Collectors` view (ignore this step if using the default): 
```xml
<cell type="subview" viewname="Collectors" id="collect" name="collectors" colspan="13" defaulttype="table" rows="3"/>
```
- Ensure the Collector subview correctly renders the default `Collector` view (pictured below) when the Subview is: 
  - [x] In Subform Mode
  - [x] In Grid Mode and the cell is expanded
  - [ ] Also ensure the correct collapsed view is shown when the Subview is in Grid Mode and the cell is collapsed
- On the CollectingEvent form definition, modify the viewname for the `collectors` relationship to be `Collector`
  - [x] Ensure the collapsed view when the Subview is in Grid Mode is using the `Collector` view definition

- [ ] On the default Agent form, ensure the GroupPerson/Members subview appears as it did in `v7.9+`

Group person subform opens as right aligned and in grid mode it originally opens as collapsed and read only and it should open expanded. 

https://github.com/specify/specify7/assets/150940840/333e6ab8-8ceb-4042-943c-f414e4eadc73

File: specifyweb/frontend/js_src/lib/components/FormParse/index.ts [hunk 3]
```diff
@@ -106,18 +111,102 @@ export const fetchView = async (
...
+    ) === 'formtable';
+
+  if (useDefaultExpanded) {
+    /**
+     * To correctly resolve this case, the non-formtable altviews and viewdefs
+     * of the table's default View are merged into the definition of the
+     * view which is supposed to represent the Collapsed/Grid view for the
+     * table
+     */
+    return fetchView(strictGetTable(view.class).view).then(
```
Comment (maxpatiiuk) [inline specifyweb/frontend/js_src/lib/components/FormParse/index.ts:170]: (optional) use await here?

**Review by maxpatiiuk — APPROVED** [review]
indeed, much cleaner!

please also add tests if possible as this is critical code

**Review by CarolineDenis — APPROVED** [review]

**Review by combs-a — APPROVED** [review]
### Testing instructions
- In the instance, make sure there are no custom views defined for Collectors
- On the view definition for the CollectingEvent, make sure the `collectors` relationship is using the `Collectors` view (ignore this step if using the default): 
```xml
<cell type="subview" viewname="Collectors" id="collect" name="collectors" colspan="13" defaulttype="table" rows="3"/>
```
- Ensure the Collector subview correctly renders the default `Collector` view (pictured below) when the Subview is: 
  - [X] In Subform Mode
  - [X] In Grid Mode and the cell is expanded
  - [X] Also ensure the correct collapsed view is shown when the Subview is in Grid Mode and the cell is collapsed
- On the CollectingEvent form definition, modify the viewname for the `collectors` relationship to be `Collector`
  - [X] Ensure the collapsed view when the Subview is in Grid Mode is using the `Collector` view definition

- [X] On the default Agent form, ensure the GroupPerson/Members subview appears as it did in `v7.9+`

Looks good, great job!

**Comment by melton-jason** [comment]
> Group person subform opens as right aligned and in grid mode it originally opens as collapsed and read only and it should open expanded.
> 
>  MU98jDWXJY.mp4

The right-aligned problem is not caused by this PR. 

Check out the default forms for `GroupPerson`:
https://github.com/specify/specify6/blob/945d28f00288d2d64979848fb2fb63fda34bdd1b/config/backstop/global.views.xml#L1095-L1117

Specifically the `columnDef` attribute: 

```xml
            <columnDef>60px,2px,p:g,2px</columnDef>
```

If this columnDef is copied to any other view definition and two labels & fields are added to two rows (like the GroupPerson default form), the same issue occurs. 

---
Any behavior with the rows being collapsed/expanded does not seem exclusive to the Collector/GroupPerson (and other affected views I am assuming). 

The behavior I am observing occurs for all SubViews: 
- If the default type of the SubView is `table`, the record becomes expanded whenever added through the plus button
- If the default type of the SubView is `form`, the all records becomes collapsed when switched to table mode, then all other records added while in table mode become expanded

Is this consistent for other SubViews in `edge` and not in this PR?

**Review by emenslin — APPROVED** [review]
### Testing instructions
- In the instance, make sure there are no custom views defined for Collectors
- On the view definition for the CollectingEvent, make sure the `collectors` relationship is using the `Collectors` view (ignore this step if using the default): 
```xml
<cell type="subview" viewname="Collectors" id="collect" name="collectors" colspan="13" defaulttype="table" rows="3"/>
```
- Ensure the Collector subview correctly renders the default `Collector` view (pictured below) when the Subview is: 
  - [x] In Subform Mode
  - [x] In Grid Mode and the cell is expanded
  - [x] Also ensure the correct collapsed view is shown when the Subview is in Grid Mode and the cell is collapsed
- On the CollectingEvent form definition, modify the viewname for the `collectors` relationship to be `Collector`
  - [x] Ensure the collapsed view when the Subview is in Grid Mode is using the `Collector` view definition

- [x] On the default Agent form, ensure the GroupPerson/Members subview appears as it did in `v7.9+`

Looks good!

**Pull request merged**